### DISCUSSION TITLE
Audio interfases - Handle slow chunk processing.

### DISCUSSION THREAD

**Discussion by arsamus** [rsta2/circle#400] [Q&A]
Hi, @rsta2. I am working on two projects based on the MiniSynth Pi class structure. The first one is a Sampler, and the other one is an alternative version of MiniSynth Pi with additional features like portamento, sustain, a 4-pole filter, antialiased oscillators, etc. However, I have an audio issue that I can't handle properly.

This issue occurs in the 'GetChunk' method when the processing time for a specific chunk ends after the next call due to an audio interface interruption. As a result, the audio streams (chunks) are processed in the wrong order until the system is rebooted. The best way to reproduce this issue is to double the number of VOICES that MiniSynth Pi can process simultaneously and attempt to play them at the same time.

Is there any way to handle this issue, such as detecting when sample processing exceeds the time allocated for processing the chunk or resetting the audio interface? I encounter this issue with the I2C interface, but I'm not sure if it also happens with PWM, USB, HDMI, etc.

Regards, Emiliano

**Comment by rsta2** [comment]
Hi Emiliano, you can display the maximum processing time of `GetChunk()` in MiniSynth Pi by adding the following line to the file *circle/Config.mk* after `configure`:

```
DEFINE += -DSHOW_STATUS
```

The displayed value must not exceed `CHUNK_SIZE / 2 / SAMPLE_RATE` with some margin. That is for example `2048 / 2 / 48000 ~= 21ms` for I2S. I did set the `VOICES_PER_CORE` parameter in MiniSynth Pi so that this condition is met for all Raspberry Pi models. Unfortunately I cannot tell you an other way of detecting the issue, you have described.

Newer synth projects (e.g. [mt32-pi](https://github.com/dwhinham/mt32-pi), [MiniDexed](https://github.com/probonopd/MiniDexed)) use the method `Write()` of the sound device object instead of overriding `GetChunk()`. This has the advantage, that the IRQ handler only needs to copy the sound data into the DMA buffer, instead of doing the entire sound processing, and this issue shouldn't occur there any more. The *sample/34-sounddevices* shows, how this can be implemented.

**Reply by arsamus** [reply]
Thank you very much for your response. I will try both alternatives you mentioned. Regarding the first one, in addition, I reduced the chunk size to the minimum possible to reduce latency. To achieve this, I had to work on minimizing multiplications, divisions, etc., as well as replacing functions from the MATH library (sin, cos, etc.) with look up tables. 

Regards! Emiliano
